tephra-core/src/test/java/co/cask/tephra/visibility/VisibilityBarrierTest.java

```diff
@@ -0,0 +1,243 @@
+ * +------+----------------------+----------------------+--------------------+--------------------+
+ * |    1 | Before Write         | Before Write         | No                 | No                 |
+ * |    2 | Before Write         | During Write         | No                 | No                 |
+ * |    3 | Before Write         | After Write          | No                 | No                 |
+ * |    4 | Before Write         | Before Write Barrier | No                 | No                 |
+ * |    5 | Before Write         | During Write Barrier | No                 | Yes                |
+ * |    6 | Before Write         | After Write Barrier  | Yes                | No                 |
+ * |      |                      |                      |                    |                    |
+ * |    7 | During Write         | During Write         | No                 | No                 |
+ * |    8 | During Write         | After Write          | No                 | No                 |
+ * |    9 | During Write         | Before Write Barrier | No                 | No                 |
+ * |   10 | During Write         | During Write Barrier | No                 | Yes                |
+ * |   11 | During Write         | After Write Barrier  | Yes                | No                 |
+ * |      |                      |                      |                    |                    |
+ * |   12 | After Write          | After Write          | No                 | No                 |
+ * |   13 | After Write          | Before Write Barrier | No                 | No                 |
+ * |   14 | After Write          | During Write Barrier | No                 | Yes #              |
+ * |   15 | After Write          | After Write Barrier  | Yes #              | No                 |
+ * |      |                      |                      |                    |                    |
+ * |   16 | Before Write Barrier | Before Write Barrier | No                 | No                 |
+ * |   17 | Before Write Barrier | During Write Barrier | No                 | Yes #              |
+ * |   18 | Before Write Barrier | After Write Barrier  | Yes #              | No                 |
+ * |      |                      |                      |                    |                    |
+ * |   19 | During Write Barrier | During Write Barrier | No                 | No                 |
+ * |   20 | During Write Barrier | After Write Barrier  | No                 | No                 |
+ * |      |                      |                      |                    |                    |
+ * |   21 | After Write Barrier  | After Write Barrier  | No                 | No                 |
+ * +------+----------------------+----------------------+--------------------+--------------------+
+ *
+ * Note: Cases marked with '#' in conflict column should not conflict, however current implementation causes
```

Comment:
This table is great. It would be good to mention again that the 'Yes' conflicts are possible using the VisibilityBarrier and that '#' false-positive conflicts are truly not conflicts but are simply side-effects of the implementation.